**Deep learning and neural networks assignment-5**

**1**. **Implement Naïve Bayes method using scikit-learn library**

   **Use dataset available with name glass**

   **Use train_test_split to create training and testing part**

   **Evaluate the model on test part using score and classification_report(y_true, y_pred)**


In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report


In [ ]:
data = pd.read_csv("glass.csv")


In [ ]:
data.shape

(214, 10)

In [101]:
data.head(5)

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [102]:
glass.tail(5)

RI     Na   Mg    Al     Si     K    Ca    Ba   Fe  Type
209  1.51623  14.14  0.0  2.88  72.61  0.08  9.18  1.06  0.0     7
210  1.51685  14.92  0.0  1.99  73.06  0.00  8.40  1.59  0.0     7
211  1.52065  14.36  0.0  2.02  73.42  0.00  8.44  1.64  0.0     7
212  1.51651  14.38  0.0  1.94  73.61  0.00  8.48  1.57  0.0     7
213  1.51711  14.23  0.0  2.08  73.36  0.00  8.62  1.67  0.0     7

In [103]:
glass.isnull().sum()

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64

In [104]:
glass.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [105]:
X = data.iloc[:, :-1]
y = data.iloc[:, -1]


In [106]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)


In [107]:
classifier = GaussianNB()


In [108]:
classifier.fit(X_train, y_train)


GaussianNB()

In [109]:
y_pred = classifier.predict(X_test)


In [110]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

print(classification_report(y_test, y_pred))


Accuracy: 0.4074074074074074
              precision    recall  f1-score   support

           1       0.62      0.26      0.37        19
           2       1.00      0.20      0.33        15
           3       0.14      0.57      0.22         7
           5       1.00      0.25      0.40         4
           6       0.75      1.00      0.86         3
           7       0.67      1.00      0.80         6

    accuracy                           0.41        54
   macro avg       0.70      0.55      0.50        54
weighted avg       0.71      0.41      0.42        54



**2. Implement linear SVM method using scikit library**
   
   **Use the same dataset above**

   **Use train_test_split to create training and testing part**

   **Evaluate the model on test part using score and**
   **classification_report(y_true, y_pred)**

In [ ]:
import numpy as np
import pandas as pd
from sklearn.svm import LinearSVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report


In [ ]:
data = pd.read_csv("glass.csv")


In [ ]:
data.shape

(214, 10)

In [ ]:
data.head(5)

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [ ]:
data.tail(5)

RI     Na   Mg    Al     Si     K    Ca    Ba   Fe  Type
209  1.51623  14.14  0.0  2.88  72.61  0.08  9.18  1.06  0.0     7
210  1.51685  14.92  0.0  1.99  73.06  0.00  8.40  1.59  0.0     7
211  1.52065  14.36  0.0  2.02  73.42  0.00  8.44  1.64  0.0     7
212  1.51651  14.38  0.0  1.94  73.61  0.00  8.48  1.57  0.0     7
213  1.51711  14.23  0.0  2.08  73.36  0.00  8.62  1.67  0.0     7

In [ ]:
glass.isnull().sum()

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64

In [ ]:
glass.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [91]:
X = data.iloc[:, :-1]
y = data.iloc[:, -1]


In [92]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)


In [93]:
classifier = LinearSVC()


In [112]:
classifier.fit(X_train, y_train)



GaussianNB()

In [113]:
y_pred = classifier.predict(X_test)


In [114]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

print(classification_report(y_test, y_pred))


Accuracy: 0.4074074074074074
              precision    recall  f1-score   support

           1       0.62      0.26      0.37        19
           2       1.00      0.20      0.33        15
           3       0.14      0.57      0.22         7
           5       1.00      0.25      0.40         4
           6       0.75      1.00      0.86         3
           7       0.67      1.00      0.80         6

    accuracy                           0.41        54
   macro avg       0.70      0.55      0.50        54
weighted avg       0.71      0.41      0.42        54



**Which algorithm you got better accuracy? Can you justify why?**

from the above algorithms, the linearsvc classifier has a slightly better accuracy than the Gaussian classifier. the accuracy of the linearscv classifier is 0.48148148148148145, and the accuracy of the gaussiannb classifier is 0.4074074074074074. This is because the linear SCV classifier is a more complex model that can learn more complex relationships between the features and the labels.
* but the gaussianNB classifier is a simpler model that is easier to interpret. it is also less likely to overfit the training data, which can lead to better performance on unseen data.
*so the best algorithm for a particular problem depends on the specific characteristics of the data and the desired trade-off between accuracy and interpretability.
*in this case the glass classification problem, the linearSCVc classifier has a slightly better accuracy than the gaussianNB classifier